v1.0 (#1) (#18)
ChainAgnostic/varsig: README.md

69227d2c202020202020202020202020202020202020202020202020202020202020202020202020
20202020202020202020202020202020202020202020202020202020202020202020202020202020
20202020202020202020202020202020202020202020202020202020202020202020202020202020
2020202020202020202020207b222f223a20226261666b72656963373579646735766b773332346f
716b636d716c74667663336b6976796e67716b69626a6f7973647769696c616b68347a356665227d
2ca202020207b222f223a20226261666b726569666664697a3672616634367a72723362327573756
6677a35666f34346167676d6f637a347a61707072366b6868686c6a63647079227da20205d2ca202
022736967223a2022387566615339773343474e386362515455536f4c31693765614b69574c53587
344324c625a566d764d397a4622a7d
```

Decoded to a string, the above reads as follows:

```
{\n
  {\n
    "role": "user",\n
    "role": "admin",\n
    "links": [\n
      {"/": "bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"},\n
      {"/": "bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"},\n
      {"/": "bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"}\n
    ],\n
  },\n
  "sig": "8ufaS9w3CGN8cbQTUSoL1i7eaKiWLSXsD2LbZVmvM9zF"\n
}
```

> [!NOTE]
> The JSON above contains a duplicate `role` key.

> Next, the application parses the JSON with the browser's native JSON parser. Only one `role` key is possible in a JavaScript object, and which one is kept is not consistency across implementations.

Reviewer: ```suggestion
Next, the application parses the JSON with the browser's native JSON parser. Only one `role` key is possible in a JavaScript object, and which one is kept is not consistent across implementations.
```

# Safety

Data already parsed to an in-memory representation can be canonically encoded trivially: it has already been through a [parser / validator][Parse Don't Validate].

Data purporting to conform to an IPLD encoding (such as [DAG-JSON]) MUST be validated prior to signature verification. This MAY be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signaled if it does not match.

> Implementers may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
>
> — [DAG-JSON Spec][DAG-JSON]

As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.

# Format

A Varsig MUST have metadata about both the [signature] and [payload encoding] that was signed over. Either field MAY be composed of one or more segments. The number of segments MUST be determined by the first segment. Recursive sub-segments MAY be used.

Varsig itself MUST contain the following segments:

* [Prefix]: The Varsig [multicodec] prefix `0x34`

> * [Version]: The Varsig version number `0x01`

Reviewer 2: I'm happy to see the addition of a version number for the `varsig` encoding.  This may very well be the wrong spot for this question but ... In this section of the [UCAN spec](https://github.com/ucan-wg/spec#envelope), the `.1.h` field description says it's a `Varsig` v1 header.  In the example JSON just below the diagram, the value `NBIFEgEAcQ` decodes to a `Varsig` v0 that's missing its signature bytes.  The `go-varsig` implementation therefore currently allows `Varsig` v0 decoding that have the correct number of signature bytes *OR* with zero signature bytes (in which case both an error is returned as well as the decoded value - ugh).  I'm hoping you'll tell us to go fix our UCAN implementation and that we can deprecate the v0 code in `go-varsig`.

Author: Yes, feel free to deprecate any Varsig v0 code. I'll update the UCAN spec to fix that prefix, and I'll merge this PR momentarily to clear up the Varsig end 👍

* [Signature Algorithm]: A signature algorithm tag and any additional fields needed to configure it
* [Payload Encoding]: The codec used to render the payload to binary

A Varsig MUST begin with one or more segments that configure the signature.

``` mermaid
block-beta
    columns 4

    Varsig:4
    prefix["Varsig Prefix\n0x34"]
    version["Version 1\n0x01"]                
    SigDetails["Signature Algorithm"]
    Encoding["Payload Encoding"]

    style Varsig fill:none;stroke:none;
```

<details>

<summary>ABNF</summary>

```abnf
varsig-header = %x34 signature-algorithm-metadata payload-encoding-metadata
signature-algorithm-metadata = unsigned-varint
payload-encoding-metadata = unsigned-varint
```

</details>

For example, an [RS256] signature over some [DAG-CBOR] is as follows:
 
```mermaid
block-beta
    block:Header
        columns 1
    
        vsig_header["Header"]
    
        block:HeaderBody
            columns 2
    
            prefix["Varsig Prefix\n0x34"]
            version["Version 1\n0x01"]
        end
    end
    
    block:Algo
        columns 1
    
        algo_header["Algorithm"]
    

>         block:AlgoBody

Reviewer 2: Looking closer, the examples in `common.md` have the same field order as the v0.1.0-rc1 documentation so I think this is just a typo in the diagram.

> Implementers may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
>
> — [DAG-JSON Spec][DAG-JSON]

As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.

## Signing CIDs

Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it is impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, storage, data availability, and networking consequences.

### Caching & Invalidation

Signing CIDs has two additional caching consequences:

1. Signing CIDs enables a simple strategy for caching validation by CID.
2. Such a strategy also MAY require accounting for revocation of the signing keys themselves. In this case, the cache would need to include additional information about the signing key.

> # Format

Reviewer 3: everything above this section IMO should be removed! IPLD, CID, etc, etc 
this spec SHOULD not care about any of it

Author: This spec is explicitly about how to sign over IPLD data. It would be called "multisig" except that's the name of an existing concept (namespace conflict). Part of the motivation is that we can't sign over IPLD data without knowing the serialisation and whatnot. If someone wants to sign non-IPLD data, they can use the raw encoding prefix.

Reviewer 3: it has that benefit of helping ipld data signatures but not restrict to it.
we talk about EIP191 and JWT in which the payload is not ipld data (or may not be).

>If someone wants to sign non-IPLD data, they can use the raw encoding prefix.

what does this mean? itsnt varsig just a bunch of prefixes?

On the other hand, the non-IPLD data schemes we want to support are strange to combine with varsig,

- JWT: i dont even know how to do JWT with varsig cause JWT is kinda similar to varsig just a json header 
- EIP191: you are supposed to sign UTF8 string using personal_sign but you can encode varsig+payload and it works
- webauthn: still need to recheck this but im guessing we will put varsig+payload in the challenge as a string or a CID

A varsig has two main segments:

* [Header]: binary coded signature metadata
* [Signature]: the signature itself

> Including the varsig header in the payload that is signed over is RECOMMENDED. Doing so eliminates any ambiguity of the signed payload format and signature algorithm configuration.

Reviewer 3: either we make it REQUIRED to include the varsig header in the signature or we need to signal inclusion in the header so its clear how to verify

Author: Added a section per our out of band conversation 👍

   Prefix           Metadata          Encoding
 ┌────┴───┐┌────────────┴───────────┐┌────┴───┐
 │        ││ ┏━━━━━━┓┏━━━━━━┓┏━━━━┓ ││ ┏━━━━┓ │
 │  0x34  ││ ┃0x1205┃┃0x0100┃┃0x12┃ ││ ┃0x71┃ │
 │        ││ ┗━━━▲━━┛┗━━━▲━━┛┗━━▲━┛ ││ ┗━━▲━┛ │
 └────────┘└─────┼───────┼──────┼───┘└────┼───┘
                 │       │      │         │
                RSA  256-bytes  │     DAG-CBOR
                    (2048 bits) │
                                │
                            SHA2-256
```

A (canonicalized) [JWT] signed with [ES256K] is as follows:

```
    Varsig
  Multiformat      Signature           Payload
    Prefix         Metadata            Encoding
  ┌────┴───┐┌───────────┴──────────┐┌─────┴────┐
  │        ││ ┏━━━━┓┏━━━━━━┓┏━━━━┓ ││ ┏━━━━━━┓ │
  │  0x34  ││ ┃0xEC┃┃0x1200┃┃0x1B┃ ││ ┃0x6A77┃ │
  │        ││ ┗━━▲━┛┗━━▲━━━┛┗━━▲━┛ ││ ┗━━━▲━━┛ │
  └────────┘└────┼─────┼───────┼───┘└─────┼────┘
                 │     │       │          │
               ECDSA   │   Keccak-256    JWT
                       │
                     P-256
```

### Varsig Prefix

The varsig prefix MUST be the [multicodec] value `0x34`.

### Signature Algorithm Metadata

> The signature algorithm field MUST consist of one or more unsigned varint segments. The signature algorithm is often the [multicodec] of the associated public key, but MAY be unique for the signature type. The code MAY live outside the multicodec table. The first segment MUST act as a discriminant for how many expected segments follow in the signature algorithm field.

Reviewer 3: ```suggestion
The signature algorithm segment MUST consist of one or more [unsigned varints](https://github.com/multiformats/unsigned-varint). 
The first `varint` in the this segment is often the [multicodec] associated with a public key, but MAY be unique for a signature algorithm and the `code` live outside of this spec and the [multicodec] table.
This first `varint` MUST also be a discriminant for the number of varints that describe the full signature algorithm.

Examples:
\```
// RSA should read 2 more varints for hash function and key size
0x1205(rsa) 0x12(sha256) 0x0100(256 bytes)

// P-256 should read one more varint for hash function
0x1200(p-256) 0x12(sha256)

// Secp256k1 should read one more varint for hash function
0xe7(secp256k1) 0x1b(keccack-256)

// Ed25519 should NOT read more varints and skip to encoding segment
0xed(ed25519)
\```
A list of common signature algorithm can be found in the [Appendix: Common Varsig Headers](/common.md)
```

> ``` abnf
> varsig-signature-algorithm
>   = %xB1   bls-public-key-curve        multihash-header ; BLS
>   / %xEC   ecdsa-curve recovery-parity multihash-header ; ECDSA
>   / %xED   eddsa-curve                 multihash-header ; EdDSA
>   / %x1205 rsa-size                    multihash-header ; RSASSA-PKCS #1 v1.5
> ```

Reviewer 3: ```suggestion
``` abnf
varsig-signature-algorithm
  = %xB1   bls-public-key-curve        multihash-header ; BLS
  / %xEC   ecdsa-curve recovery-parity multihash-header ; ECDSA
  / %xED   eddsa-curve                 multihash-header ; EdDSA
  / %x1205 rsa-size                    multihash-header ; RSASSA-PKCS #1 v1.5
```
```

### Payload Encoding Metadata

> The [IPLD] data model is encoding agnostic by design. This is very convenient in many applications, such as making for very convenient conversions between types for transmission versus encoding. Unfortunately signatures require signing over specific bytes, and thus over a specific encoding of the data. To facilitate this, the type `varsig-encoding-info` MUST be used:

Reviewer 3: we should remove all mention to IPLD from this spec it shouldnt care if its ipld data or not

Author: ```
## Dependencies

* [IPLD]
* [Multibase]
* [Multicodec]

# Abstract

Varsig is a [multiformat][Multiformats] for describing signatures over IPLD data and raw bytes in a way that preserves information about the payload and canonicalization metadata.
```

> The [IPLD] data model is encoding agnostic by design. This is very convenient in many applications, such as making for very convenient conversions between types for transmission versus encoding. Unfortunately signatures require signing over specific bytes, and thus over a specific encoding of the data. To facilitate this, the type `varsig-encoding-info` MUST be used:

Reviewer 3: ```suggestion
The [IPLD] data model is encoding agnostic by design. This is very convenient in many applications, such as making for very convenient conversions between types for transmission versus encoding. Unfortunately signatures require signing over specific bytes, and thus over a specific encoding of the data. To facilitate this, the type `varsig-encoding-metadata` MUST be used:
```

``` abnf

> varsig-encoding-info

Reviewer 3: ```suggestion
varsig-encoding-metadata
```

  = %x5F                        ; Single verbatim payload (without key)
  / %x70                        ; DAG-PB multicodec prefix
  / %x71                        ; DAG-CBOR multicodec prefix
  / %x0129                      ; DAG-JSON multicodec prefix
  / %x6A77                      ; Canonicalized JWT

>   / %xE191 varsig-encoding-info ; EIP-191 "personal sign"

Reviewer 3: ```suggestion
  / 0xd191 ; EIP-191 "personal sign"
```

Author: I think we still need to know how to format the payload, right? IIRC the EIP-191 just sticks the domain separator to the fully serialized payload, or is there some detail I'm forgetting?

Reviewer 3: ah i see what you mean, you are right! we may even need to put encoding and serialization.
need to check if wallets support sending bytes or only strings

``` abnf
varsig-signature-algorithm
  = %xB1   bls-public-key-curve        multihash-header ; BLS
  / %xEC   ecdsa-curve recovery-parity multihash-header ; ECDSA
  / %xED   eddsa-curve                 multihash-header ; EdDSA
  / %x1205 rsa-size                    multihash-header ; RSASSA-PKCS #1 v1.5
```

### Payload Encoding Metadata

The [IPLD] data model is encoding agnostic by design. This is very convenient in many applications, such as making for very convenient conversions between types for transmission versus encoding. Unfortunately signatures require signing over specific bytes, and thus over a specific encoding of the data. To facilitate this, the type `varsig-encoding-metadata` MUST be used:

``` abnf
varsig-encoding-metadata
  = %x5F                        ; Single verbatim payload (without key)
  / %x70                        ; DAG-PB multicodec prefix
  / %x71                        ; DAG-CBOR multicodec prefix
  / %x0129                      ; DAG-JSON multicodec prefix
  / %x6A77                      ; Canonicalized JWT
  / %xE191 varsig-encoding-info ; EIP-191 "personal sign"
```

## Signature Bytes

> The associated signature bytes MUST be represented as a byte array. The MAY come directly after the varsig, but use of the [Varsig Envelope] is RECOMMENDED.

Reviewer 3: ```suggestion
The associated signature bytes MUST be represented as a byte array. The signature bytes MAY come directly after the varsig, but use of the [Varsig Envelope] is RECOMMENDED.
```
